src/Stratis.Bitcoin.Features.PoA/BasePoAFeatureConsensusRules/PoAHeaderSignatureRule.cs

```diff
@@ -71,6 +76,42 @@ public override void Run(RuleContext context)
                     }
                 }
 
+                try 
+                {
+                    // Gather all past and present mining public keys.
+                    IEnumerable<PubKey> genesisFederation = ((PoAConsensusOptions)this.network.Consensus.Options).GenesisFederationMembers.Select(m => m.PubKey);
+                    var knownKeys = new HashSet<PubKey>(genesisFederation);
+                    foreach (Poll poll in this.votingManager.GetFinishedPolls().Where(x => ((x.VotingData.Key == VoteKey.AddFederationMember) || (x.VotingData.Key == VoteKey.KickFederationMember))))
+                    {
+                        IFederationMember federationMember = ((PoAConsensusFactory)(this.network.Consensus.ConsensusFactory)).DeserializeFederationMember(poll.VotingData.Data);
+                        knownKeys.Add(federationMember.PubKey);
+                    }
+
+                    // Try to provide the public key that signed the block.
+                    var signature = ECDSASignature.FromDER(header.BlockSignature.Signature);
+                    for (int recId = 0; ; recId++)
```

Comment:
I think this can only be one of a few possible values. The reason why it has to be looped is that you don't necessarily know from an arbitrary signature what recId was when it was originally signed. This is solved elsewhere by prepending the signature with header information containing the value of recId, but I don't know offhand if the block signatures have this. I suspect not. In any case an ECDSASignature instance does not have this information available by itself.

I think the loop should be constrained to looping from 0 to 3.

Reply:
We seem to be getting `null` once the `recId` goes beyond the candidate keys that are available so I avoided hard-coding a loop constant.